### ISSUE TITLE
Warn if plugin name in config is not a keyword

### ISSUE THREAD

**Issue by plexus** [lambdaisland/kaocha#36]
Plugin names are assumed to be keywords (it's an identifier for the plugin). However it's easy to mix this up and put in a symbol. We could allow this and coerce, but I'd rather enforce systematic use of keywords. That also has the benefit that symbols are still available for some later extension.

This isn't currently caught by spec I think because we let plugins run their config hooks first before validating the config, but maybe we should really validate twice, once before and once after the config hook.

**Comment by riyasingh1004** [comment]
Hi , would like to work on this issue. Please help me get started with it.

**Comment by riyasingh1004** [comment]
I will like to work on this issue. Please help me get started.

**Comment by plexus** [comment]
Hi @riyasingh1004 :wave: great to have you here!

I had a better look at what needs to happen here. There are two things that are missing.

1. We need to add a spec for `:kaocha/plugins`
2. We need to validate the configuration

Step one is actually very straightforward, in `src/kaocha/specs.clj`, we need something like this. You can add it at the top somewhere, for instance around line 14.

``` clojure
(s/def :kaocha/plugins (s/coll-of keyword?))
```

What this does is tell clojure.spec that if a map contains a `:kaocha/plugins` key, then the value corresponding with that key has to be a collection of keywords. If you are not familiar yet with clojure spec then I recommend watching [this introduction video](https://lambdaisland.com/episodes/clojure-spec) and trying it out. Ping me if you have questions.

So that is step one. We already have a spec saying that a kaocha configuration needs to contain `:kaocha/plugins` so that's good.

The second step is to actually validate the configuration so the user gets feedback. You can use [expound](https://github.com/bhb/expound) to create really friendly and helpful messages, this is what it would look like:

```
 % bin/kaocha                                                                                                                                                                                                                                                                                   ✘ 252 
Invalid configuration in tests.edn
-- Spec failed --------------------

  {:kaocha/reporter ...,
   :kaocha/color? ...,
   :kaocha/fail-fast? ...,
   :kaocha/plugins [... ... ... foo.bar ...],
                                ^^^^^^^
   :kaocha/tests ...,
   :kaocha.hooks/pre-run ...}

should satisfy

  keyword?

-- Relevant specs -------

:kaocha/plugins:
  (clojure.spec.alpha/coll-of clojure.core/keyword?)
:kaocha/config:
  (clojure.spec.alpha/keys :req ~(conj global-opts :kaocha/tests))

-------------------------
Detected 1 error
```

If you look at the bottom of `src/kaocha/specs.clj` you see that we actually already have a `assert-spec` function. This will check a certain value against a certain spec, and if it fails it gives the user feedback and exits. It looks like this

```
(defn assert-spec [spec value]
  (when-not (s/valid? spec value)
    (throw (AssertionError. (expound/expound-str spec value)))))
```

I've been thinking there are ways we could improve this, but for now let's use it as-is. We want to do the actual check after loading the configuration, but before loading the plugins. For this have a look at `kaocha.runner/-main` (`src/kaocha/runner.clj` line 141).

At lines 148-152 we load the config from tests.edn, and then based on the config we load the plugins.

``` clojure
(let [,,,
      ;; load the config
      config (-> config-file
                 (or "tests.edn")
                 (config/load-config (if profile
                                       {:profile profile}
                                       {})))

      ;; <---

      ;; load the plugins
      plugin-chain (plugin/load-all (concat (:kaocha/plugins config) plugin))
      ,,,
      ])
```

In between where I put the arrow is where we want to put the `assert-spec` check. The only tricky thing is that we're in the middle of a let block there, which is an inconvenient place to add code like this that just performs some checks and prints some stuff but doesn't actually return a value. There are two ways to deal with this, you can split the let block up.

```
(let [,,,
      config (-> config-file
                 (or "tests.edn")
                 (config/load-config (if profile
                                       {:profile profile}
                                       {})))]
  ;; <---
  (let [plugin-chain (plugin/load-all (concat (:kaocha/plugins config) plugin))
        ,,,
        ]))
```

Or you can indicate that `assert-spec` doesn't have a return value (or that you want to ignore the return value), by using `_` as the variable name on the left-hand side of the let. This is probably what I would do.

```
(let [,,,
      config (-> config-file
                 (or "tests.edn")
                 (config/load-config (if profile
                                       {:profile profile}
                                       {})))
      _ (assert-spec :kaocha/config config)
      plugin-chain (plugin/load-all (concat (:kaocha/plugins config) plugin))
      ,,,]
  ,,,)
```

See if you can get that far and if you can verify that it works. At this point the output will look something like this:

```
 % bin/kaocha                                
Exception in thread "main" java.lang.AssertionError: -- Spec failed --------------------
                                          
  {:kaocha/reporter ...,                
   :kaocha/color? ...,                            
   :kaocha/fail-fast? ...,                
   :kaocha/plugins [... ... ... foo.bar ...],
                                ^^^^^^^                                                                                                                                                                                                                                                               
   :kaocha/tests ...,                
   :kaocha.hooks/pre-run ...}    
                                    
should satisfy               
                                                             
  keyword?    
                 
-- Relevant specs -------                                    
                                                      
:kaocha/plugins:                                             
  (clojure.spec.alpha/coll-of clojure.core/keyword?)   
:kaocha/config:                                       
  (clojure.spec.alpha/keys :req ~(conj global-opts :kaocha/tests))
                                                             
-------------------------
Detected 1 error
                                                          
        at kaocha.specs$assert_spec.invokeStatic(specs.clj:103)
        at kaocha.specs$assert_spec.invoke(specs.clj:101)        
        at kaocha.runner$_main_STAR_.invokeStatic(runner.clj:153)
        at kaocha.runner$_main_STAR_.doInvoke(runner.clj:141)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.core$apply.invokeStatic(core.clj:665)
        at clojure.core$apply.invoke(core.clj:660)
        at kaocha.runner$_main.invokeStatic(runner.clj:174)
        at kaocha.runner$_main.doInvoke(runner.clj:172)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.core$apply.invokeStatic(core.clj:665)
        at clojure.main$main_opt.invokeStatic(main.clj:491)
        at clojure.main$main_opt.invoke(main.clj:487)
        at clojure.main$main.invokeStatic(main.clj:598)
        at clojure.main$main.doInvoke(main.clj:561)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.main.main(main.java:37)
```

I think the stack trace in this case doesn't add much information, so we can still make this better, but let's see if you can get this far first.

**Comment by plexus** [comment]
Hi @riyasingh1004 , I'm going to unassign you so others can work on this. Do let me know if you are still interested in trying this or other issues.

**Comment by alysbrooks** [comment]
Fixed by #169 🎉
